### PR TITLE
CDD 1459

### PR DESCRIPTION
**Pull request by paulblakehmrc**
Implementation of /send-entry-usage endpoint

### CODE REVIEW

File: app/uk/gov/hmrc/customs/dit/licence/controllers/EntryUsageController.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                                      logger: LicencesLogger) extends BaseController {
+
+  private val configKey = "dit-lite-entry-usage"
+
+  def post(): Action[AnyContent] = (Action andThen validateAndExtractHeadersAction).async(bodyParser = xmlOrEmptyBody) {
+      implicit validatedRequest: ValidatedRequest[AnyContent] =>
+
+      logger.info("entered EntryUsageController after validating headers")
+      validatedRequest.request.body.asXml match {
+        case Some(_) =>
```
Comment (googley42) [inline app/uk/gov/hmrc/customs/dit/licence/controllers/EntryUsageController.scala]: what if connector returns failed future - should we be handling this in a more graceful way together with logging?

File: test/unit/connector/DitLiteConnectorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    }
+
+    "when making an failing request" should {
+      "propagate an underlying error when DIT-LITE call fails with a non-http exception" in {
+        returnResponseForRequest(Future.failed(emulatedServiceFailure))
+
+        val caught = intercept[EmulatedServiceFailure] {
+          awaitRequest
+        }
```
Comment (googley42) [inline test/unit/connector/DitLiteConnectorSpec.scala]: blank line after `WHEN` block pls

File: test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  val blockReturningOk = (_: ValidatedRequest[_]) => Future.successful(Ok)
+
+  private val mockHeaderValidator = mock[HeaderValidator]
+  private val mockLogger = mock[LicencesLogger]
+  private val actionRefiner = new ValidateAndExtractHeadersAction(mockHeaderValidator, mockLogger)
+
+  "ValidateAndExtractHeadersAction" should {
+    "when valid headers are submitted then return a ValidatedRequest" in {
+      val request = FakeRequest().withXmlBody(ValidXML).withHeaders(ValidHeaders.toSeq:_*)
+
```
Comment (googley42) [inline test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala]: no blank lines between stuff in `GIVEN` block pls

File: test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      when(mockHeaderValidator.validateHeaders(request)).thenReturn(Right(TestRequestData))
+
+      val result: Result = await(actionRefiner.invokeBlock(request, blockReturningOk))
+
+      result shouldBe Ok
+    }
+
+    "when invalid headers are submitted then return an ErrorResponse" in {
+      val request = FakeRequest().withXmlBody(ValidXML).withHeaders(InvalidHeaders.toSeq:_*)
+
```
Comment (googley42) [inline test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala]: no blank lines between stuff in `GIVEN` block pls

File: test/util/TestData.scala [hunk 5]
```diff
@@ -16,21 +16,67 @@
...
+
+  lazy val MalformedXmlRequest: FakeRequest[AnyContentAsText] = ValidRequest.withTextBody("<xml><non_well_formed><xml>")
+
+  lazy val ErrorXCorrelationIdMissingOrInvalid = errorBadRequest("X-Correlation-ID is missing or invalid")
+  lazy val ErrorUnauthorizedBasicToken = ErrorResponse(UNAUTHORIZED, UnauthorizedCode, "Basic token is missing or not authorized")
 
+  val TestRequestData = RequestData(correlationId)
 }
+
 object RequestHeaders {
```
Comment (googley42) [inline test/util/TestData.scala]: public constants should be `CamelCase`

File: test/unit/logging/LoggingHelperSpec.scala [hunk 6]
```diff
@@ -16,73 +16,28 @@
...
+import play.api.mvc.{AnyContent, Request}
 import uk.gov.hmrc.customs.dit.licence.logging.LoggingHelper
-import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.customs.dit.licence.model.{RequestData, ValidatedRequest}
 import uk.gov.hmrc.play.test.UnitSpec
+import util.TestData.correlationId
 
-import util.RequestHeaders._
-import util.TestData.correlationIdValue
+class LoggingHelperSpec extends UnitSpec with MockitoSugar {
```
Comment (googley42) [inline test/unit/logging/LoggingHelperSpec.scala]: unit tests with mocks/state that need to be reset should use either use ScalaTest lifecycle methods like `beforeEach` or a simpler more robust `SetUp` trait pattern - which is instantiated for each test. Applies to all unit tests.

File: app/uk/gov/hmrc/customs/dit/licence/controllers/HeaderValidator.scala [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    def leftWithLogContainingValue(s: String) = {
+      logger.error(s"${errorResponse.message} '$s'")
+      left
+    }
+
+    h.get(headerName).fold[Either[ErrorResponse, String]]{
+      leftWithLog
+    }{
+      h => if (rule(h)) Right(h) else leftWithLogContainingValue(h)
```
[the remark this answers is not in the document]
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/customs/dit/licence/controllers/HeaderValidator.scala]: Done

File: README.md [hunk 8]
```diff
@@ -2,7 +2,89 @@
...
+Two endpoints are available. Both require an authorisation token.
+
+### Licence Usage
+
+Endpoint - /send-entry-usage
+
+
+### Late Licence Usage
+
+Endpoint - /send-late-usage
```
[the remark this answers is not in the document]
Comment (paulblakehmrc) [reply README.md]: Done

File: app/uk/gov/hmrc/customs/dit/licence/controllers/EntryUsageController.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+          ditLiteConnector.post(configKey).map { response =>
```
[the remark this answers is not in the document]
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/customs/dit/licence/controllers/EntryUsageController.scala]: Here I'm converting `response.allHeaders`, a `Map[String, scala.Seq[String]]` into a type suitable for creating a `ResponseHeader` where `Map[String, String]` is required. I've extracted it to a method as I can't remove to completely.

File: test/unit/connector/DitLiteConnectorSpec.scala (hunk 2 above)
Comment (paulblakehmrc) [reply test/unit/connector/DitLiteConnectorSpec.scala]: Done

File: test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala (hunk 3 above)
Comment (paulblakehmrc) [reply test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala]: Done

File: test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala (hunk 4 above)
Comment (paulblakehmrc) [reply test/unit/controllers/ValidateAndExtractHeadersActionSpec.scala]: Done

File: test/util/TestData.scala (hunk 5 above)
Comment (paulblakehmrc) [reply test/util/TestData.scala]: Done

File: test/unit/logging/LoggingHelperSpec.scala (hunk 6 above)
Comment (paulblakehmrc) [reply test/unit/logging/LoggingHelperSpec.scala]: Done

File: app/uk/gov/hmrc/customs/dit/licence/connectors/DitLiteConnector.scala [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+@Singleton
+class DitLiteConnector @Inject()(wsHttp: WSHttp,
+                                 serviceConfigProvider: ServiceConfigProvider,
+                                 logger: LicencesLogger) extends RawResponseReads {
+
+  def post(configKey: String)(implicit validatedRequest: ValidatedRequest[AnyContent]): Future[HttpResponse] = {
+
```
[the remark this answers is not in the document]
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/customs/dit/licence/connectors/DitLiteConnector.scala]: One test added and the need for the another removed

**Pull request merged**